### ISSUE TITLE
Check for update after rollback doesn't notify version changed

### ISSUE THREAD

**Issue by jHabjanMXP** [angular/angular#46644]
### Which @angular/* package(s) are the source of the bug?

service-worker

### Is this a regression?

No

### Description

Relates to #42940

Steps to reproduce:
- Step 1: 
  - Install version V1, open webpage
- Step 2:
  - Install version V2
  - call `swUpdate.checkForUpdate()`
  - get notified about version change by `swUpdate.versionUpdates`
  - after prompt call `swUpdate.activateUpdate()` 
  - and then `document.location.reload()`
- Step 3:
  - install version V1
  - call `swUpdate.checkForUpdate()`

**Result**: we don't get notified about version change by versionUpdates subscription

Here is the reason:

```javascript
async checkForUpdate() {
  let hash = "(unknown)";
  try {
    const manifest = await this.fetchLatestManifest(true);
    if (manifest === null) {
      this.debugger.log("Check for update aborted. (Client or server offline.)");
      return false;
    }
    hash = hashManifest(manifest);
    // if (this.versions.has(hash) { <<<<<<<<<<<<<<<<<<< current detection which doesn't check about rolled-back hash
    if (this.versions.has(hash) && hash === this.latestHash) { <<<<<<<< I added check to against latest hash
      return false;
    }
    await this.notifyClientsAboutVersionDetected(manifest, hash);
    await this.setupUpdate(manifest, hash);
    return true;
  } catch (err) {
    this.debugger.log(err, `Error occurred while updating to manifest ${hash}`);
    this.state = DriverReadyState.EXISTING_CLIENTS_ONLY;
    this.stateMessage = `Degraded due to failed initialization: ${errorToString(err)}`;
    return false;
  }
}
```

From the #42940 thread, I still don't get the solution in our case.
New app detection only works when re-open the browser. I guess `cleanupCaches()` is called by `initialize` which then takes latest hash into account

### Please provide the exception or error you saw

```true
Application isn't notified about downgraded version if browser isn't re-opened, if we install V1, install V2, then downgrade to V1 (all in timespan when browser is opened)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 13.0.3
Node: 16.15.1
Package Manager: npm 8.5.5
OS: win32 x64

Angular:
...

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.1300.3 (cli-only)
@angular-devkit/core         13.0.3 (cli-only)
@angular-devkit/schematics   13.0.3 (cli-only)
@schematics/angular          13.0.3 (cli-only)
```

**Comment by gkalpak** [comment]
@jHabjanMXP, I am not sure what this issue is about or what you are asking.

AFAICT, this issue has been discussed in #42940 and #24338. Are you saying that the explanations there were not clear? Or are you disagreeing with the current approach? (If so, what are you suggesting?)

**Comment by jHabjanMXP** [comment]
@gkalpak As I see, current workaround is to change anything in `ngsw-config.json` and then new sw is registered. But our approach is to use the same binaries when rolling back, without making manual changes to `ngsw-config.json`

If there's no clean solution for this "workaround", it's fine by me, but I just wanted to bring this topic up again, so we'll know how to handle this

**Comment by gkalpak** [comment]
That's right. There is currently no way to force the SW to update to a version if it has the same version (i.e. same binaries) cached and marked as not being the latest version. The reason is that it is not possible for the SW to distinguish rolling back to an earlier version vs a cache/proxy serving an outdated version.

The only solution I can think of would be to have a time threshold after which we would consider a version fetched from the server as a new one (even if its identical to a still cached older version). We could even make it configurable.
For example, if the SW installed version `A`, then a day later installed version `B` and then saw version `A` again, since a day has passed since it first installed it, it would treat it as a new version.

The problem here is that this would not be 100% reliable either (unless you made the threshold really small), because a client may have installed version `A` for the first time right before installing version `B` (and not when version `A` was first deployed to the server).

##
Given that (a) there is no reliable solution for this, (b) any mitigation we would implement would make the SW behavior less predictable and (c) there hasn't been a big demand for this, I lean towards closing this issue as "won't fix".

WDYT, @jHabjanMXP?

**Comment by jHabjanMXP** [comment]
@gkalpak Thank you for the explanation. Can be closed
